content/blog/2024/10/policy-groups.md

```diff
@@ -0,0 +1,276 @@
+# my-service.yml
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  selector:
+    app.kubernetes.io/name: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+EOF
+
+service/my-service created
+```
+
+As expected, the service creation succeeded. It is the first service that uses the selector for `MyApp`.
+
+Now, if we create a second service, reusing the selector:
+
+```console
+kubectl apply -f - <<EOF
+# second-service.yml
+apiVersion: v1
+kind: Service
+metadata:
+  name: second-service
+spec:
+  selector:
+    app.kubernetes.io/name: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+EOF
+
+Error from server: error when creating "STDIN":
+admission webhook "clusterwide-group-unique-service-selector.kubewarden.admission" denied the request:
+the service selector is not unique or the service is not owned by the Quokkas team
+```
+
+We get a rejection as expected.
+
+## Obtaining full details on results
+
+We know that for rejecting that `second-service`, the Policy Group evaluated
+each of their policies, only if it was needed. Kubewarden exposes this
+information as part of the AdmissionResponse's `.status.details.causes`. This
+will include skipping policy evaluations when possible.
+
+These can be seen by tools, for example by increasing the verbosity level of kubectl:
+
+```console
+kubectl -v4 apply -f second-service.yml
+I0927 14:00:37.262003  107266 cert_rotation.go:137] Starting client certificate rotation controller
+I0927 14:00:37.527613  107266 helpers.go:246] server response object: [{
+  "kind": "Status",
+  "apiVersion": "v1",
+  "metadata": {},
+  "status": "Failure",
+  "message": "error when creating \"second-service.yml\": admission webhook \"clusterwide-group-unique-service-selector.kubewarden.admission\" denied the request: the service selector is not unique or the service is not owned by the Quokkas team",
+  "details": {
+    "causes": [
+      {
+        "message": "service is using selector(s) already defined by these services: [\"my-service\"]",
+        "field": "spec.policies.unique_service_selector"
+      },
+      {
+        "message": "The following mandatory annotations are missing: owner",
+        "field": "spec.policies.owned_by_quokkas_team"
+      }
+    ]
+  },
+  "code": 400
+}]
+Error from server: error when creating "second-service.yml": admission webhook "clusterwide-group-unique-service-selector.kubewarden.admission" denied the request: the service selector is not unique or the service is not owned by the Quokkas team
+```
+
+### Audit scanner
+
+Another way of obtaining details on the state of the cluster is the [Audit
+Scanner](https://docs.kubewarden.io/next/explanations/audit-scanner) feature.
+Policy Groups are fully supported in the Kubewarden stack, and their results
+get published to Policy Reports.
+
+## Recap
+
+We have seen the 2 new Policy Groups Custom Resources: **\*AdmissionPolicyGroups** and
```

Comment:
Did you want that asterisk in there?

Reply:
Nope, good catch! Fixed with https://github.com/kubewarden/kubewarden.io/pull/266/commits/85bfbba5feb64d8f20b4dcb8aba1f14e024e075a.